File: lib/features/favorites/data/favorites_repository.dart
import 'package:drift/drift.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../data/database/app_database.dart';

/// Stable entity-kind keys for the generic [Favorites] table. Kept as string
/// constants (not an enum) so persisted/exported rows stay valid even if the
/// set of kinds grows later.
abstract final class FavoriteKind {
  static const job = 'job';
  static const kbArticle = 'kb_article';
  static const fishingZone = 'fishing_zone';
  static const trackedObject = 'tracked_object';
}

/// Reactive store for star/pin/bookmark across entity kinds. All reads go
/// through drift `watch()` so the UI updates live as favorites toggle.
class FavoritesRepository {
  FavoritesRepository(this._db);
  final AppDatabase _db;

  /// Live set of favorited ids for a single [kind].
  Stream<Set<String>> watchIds(String kind) {
    final q = _db.select(_db.favorites)
      ..where((f) => f.entityType.equals(kind));
    return q.watch().map((rows) => rows.map((r) => r.entityId).toSet());
  }

  /// Live favorite flag for one entity.
  Stream<bool> watchIsFavorite(String kind, String id) {
    final q = _db.select(_db.favorites)
      ..where((f) => f.entityType.equals(kind) & f.entityId.equals(id));
    return q.watch().map((rows) => rows.isNotEmpty);
  }

  Future<bool> isFavorite(String kind, String id) async {
    final row = await (_db.select(_db.favorites)
          ..where((f) => f.entityType.equals(kind) & f.entityId.equals(id)))
        .getSingleOrNull();
    return row != null;
  }

  /// Flip the favorite flag for one entity. Returns the new state.
  Future<bool> toggle(String kind, String id) async {
    if (await isFavorite(kind, id)) {
      await (_db.delete(_db.favorites)
            ..where((f) => f.entityType.equals(kind) & f.entityId.equals(id)))
          .go();
      return false;
    }
    await _db.into(_db.favorites).insert(
          FavoritesCompanion.insert(
            entityType: kind,
            entityId: id,
            createdAt: DateTime.now(),
          ),
          mode: InsertMode.insertOrReplace,
        );
    return true;
  }
}

final favoritesRepositoryProvider = Provider<FavoritesRepository>((ref) {
  return FavoritesRepository(ref.watch(appDatabaseProvider));
});

/// Live favorited-id set for a given [FavoriteKind]. Family key is the kind.
final favoriteIdsProvider =
    StreamProvider.family<Set<String>, String>((ref, kind) {
  return ref.watch(favoritesRepositoryProvider).watchIds(kind);
});

File: webapp/src/features/favorites/favoritesRepository.ts
import { liveQuery, type Observable } from 'dexie';
import { db, type FavoriteKindValue } from '../../data/db';

/**
 * Generic star / pin / bookmark store shared across the whole app
 * (knowledge spec §10). Persisted to the Dexie `favorites` table with the
 * compound primary key `[entityType+entityId]`, so an entity is favorited at
 * most once. This area only ever writes `kb_article` rows, but the store is
 * shared (jobs star, fishing-zone star, tracker pin, map / zone favorites).
 *
 * All reads are live: `liveQuery` re-emits whenever the `favorites` table
 * changes, so toggling from anywhere updates every observer.
 */

async function idsForKind(kind: FavoriteKindValue): Promise<Set<string>> {
  const rows = await db.favorites.where('entityType').equals(kind).toArray();
  return new Set(rows.map((row) => row.entityId));
}

export const favoritesRepository = {
  /** Live set of favorited ids for one kind. */
  watchIds(kind: FavoriteKindValue): Observable<Set<string>> {
    return liveQuery(() => idsForKind(kind));
  },

  /** Live boolean for a single entity. */
  watchIsFavorite(kind: FavoriteKindValue, id: string): Observable<boolean> {
    return liveQuery(async () => (await db.favorites.get([kind, id])) !== undefined);
  },

  /** Plain querier for `useLiveQuery` consumers (FavoriteButton, jobs filter). */
  getIds(kind: FavoriteKindValue): Promise<Set<string>> {
    return idsForKind(kind);
  },

  async isFavorite(kind: FavoriteKindValue, id: string): Promise<boolean> {
    return (await db.favorites.get([kind, id])) !== undefined;
  },

  /**
   * Delete the row if present, else insert it (createdAt = now).
   * Returns the new favorite state (`true` = now favorited).
   */
  async toggle(kind: FavoriteKindValue, id: string): Promise<boolean> {
    return db.transaction('rw', db.favorites, async () => {
      const existing = await db.favorites.get([kind, id]);
      if (existing !== undefined) {
        await db.favorites.delete([kind, id]);
        return false;
      }
      await db.favorites.put({ entityType: kind, entityId: id, createdAt: Date.now() });
      return true;
    });
  },
};
